File: runtime/swift/prompty-model/Sources/PromptyModel/Discovery.swift
import Foundation

public enum Discovery {
  public static func enrich(_ base: ModelInfo, provider: String) -> ModelInfo {
    guard let entry = CapabilityDataset.shared.match(modelId: base.id, provider: provider) else {
      return base
    }

    var info = base
    if info.contextWindow == nil, let contextWindow = entry.contextWindow {
      info.contextWindow = contextWindow
    }
    if info.inputModalities == nil, let inputModalities = entry.inputModalities {
      info.inputModalities = inputModalities
    }
    if info.outputModalities == nil, let outputModalities = entry.outputModalities {
      info.outputModalities = outputModalities
    }
    return info
  }

  public static func mapModel(_ raw: Any?, provider: String) throws -> ModelInfo {
    guard let raw else {
      throw TypraRuntimeError.invalidObject(type: "ModelInfo")
    }
    return try mapModel(raw, provider: provider)
  }

  public static func mapModel(_ raw: Any, provider: String) throws -> ModelInfo {
    let data = try TypraRuntime.object(raw, typeName: "ModelInfo")
    var info = ModelInfo()
    info.additionalProperties = data

    switch provider {
    case "anthropic":
      info.id = string(data["id"]) ?? ""
      info.displayName = string(data["display_name"])
      info.ownedBy = "anthropic"
      info.contextWindow = int32(data["context_length"])
      info.inputModalities = stringArray(data["input_modalities"])
      info.outputModalities = stringArray(data["output_modalities"])
    case "foundry":
      mapFoundryModel(data, into: &info)
    default:
      info.id = string(data["id"]) ?? ""
      info.ownedBy = string(data["owned_by"])
    }

    return info
  }

  private static func mapFoundryModel(_ data: [String: Any], into info: inout ModelInfo) {
    if let properties = data["properties"] as? [String: Any] {
      let model = properties["model"] as? [String: Any]
      let capabilities = properties["capabilities"] as? [String: Any]
      info.id = string(data["name"]) ?? ""
      info.displayName = string(model?["name"])
      info.ownedBy = string(model?["publisher"])
      info.contextWindow = int32(model?["maxContextLength"])
      info.inputModalities = stringArray(capabilities?["supportedInputModalities"])
      info.outputModalities = stringArray(capabilities?["supportedOutputModalities"])
    } else if data["modelName"] != nil || string(data["type"]) == "ModelDeployment" {
      info.id = string(data["name"]) ?? ""
      info.displayName = string(data["modelName"])
      info.ownedBy = string(data["modelPublisher"])
      info.contextWindow = int32(data["maxContextLength"])
    } else {
      info.id = string(data["id"]) ?? ""
      info.ownedBy = string(data["owned_by"])
      info.contextWindow = int32(data["maxContextLength"])
    }
  }
}

public func enrich(_ base: ModelInfo, provider: String) -> ModelInfo {
  Discovery.enrich(base, provider: provider)
}

public func mapModel(_ raw: Any, provider: String) throws -> ModelInfo {
  try Discovery.mapModel(raw, provider: provider)
}

private struct CapabilityEntry {
  let prefix: String
  let contextWindow: Int32?
  let inputModalities: [String]?
  let outputModalities: [String]?

  init?(_ data: [String: Any]) {
    guard let prefix = string(data["prefix"]) else {
      return nil
    }
    self.prefix = prefix
    contextWindow = int32(data["contextWindow"])
    inputModalities = stringArray(data["inputModalities"])
    outputModalities = stringArray(data["outputModalities"])
  }
}

private struct CapabilityDataset {
  static let shared = CapabilityDataset.load()

  let providers: [String: [CapabilityEntry]]

  func match(modelId: String, provider: String) -> CapabilityEntry? {
    guard let entries = providers[provider] else {
      return nil
    }

    return entries
      .filter { isBoundaryPrefix($0.prefix, of: modelId) }
      .max { $0.prefix.count < $1.prefix.count }
  }

  private func isBoundaryPrefix(_ prefix: String, of modelId: String) -> Bool {
    if modelId == prefix {
      return true
    }
    guard modelId.hasPrefix(prefix), let boundary = modelId.dropFirst(prefix.count).unicodeScalars.first else {
      return false
    }
    return !CharacterSet.alphanumerics.contains(boundary)
  }

  private static func load() -> CapabilityDataset {
    guard
      let url = Bundle.module.url(forResource: "model_capabilities", withExtension: "json"),
      let data = try? Data(contentsOf: url),
      let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
      let providersObject = object["providers"] as? [String: Any]
    else {
      return CapabilityDataset(providers: [:])
    }

    var providers: [String: [CapabilityEntry]] = [:]
    for (name, rawEntries) in providersObject {
      guard let entryObjects = rawEntries as? [[String: Any]] else {
        continue
      }
      providers[name] = entryObjects.compactMap(CapabilityEntry.init)
    }
    return CapabilityDataset(providers: providers)
  }
}

private func string(_ value: Any?) -> String? {
  value as? String
}

private func int32(_ value: Any?) -> Int32? {
  if let value = value as? Int32 {
    return value
  }
  if let value = value as? Int {
    return Int32(exactly: value)
  }
  if let value = value as? Int64 {
    return Int32(exactly: value)
  }
  if let value = value as? NSNumber, !TypraRuntime.isBoolNumber(value) {
    return Int32(exactly: value.int64Value)
  }
  return nil
}

private func stringArray(_ value: Any?) -> [String]? {
  guard let array = value as? [Any] else {
    return nil
  }
  return array.compactMap { $0 as? String }
}

File: runtime/csharp/Prompty.Core/Discovery.cs
#nullable enable

using System.Reflection;
using System.Text.Json.Nodes;

namespace Prompty.Core;

public static class Discovery
{
    private const string CapabilityResourceName = "Prompty.Core.Data.model_capabilities.json";

    private static readonly Lazy<JsonObject> CapabilityData = new(LoadCapabilityData);

    public static ModelInfo Enrich(ModelInfo @base, string provider)
    {
        var entry = MatchCapabilities(@base.Id, provider);
        if (entry is null)
        {
            return @base;
        }

        if (@base.ContextWindow is null && entry.TryGetPropertyValue("contextWindow", out var contextWindow))
        {
            @base.ContextWindow = contextWindow?.GetValue<int>();
        }

        if (@base.InputModalities is null && entry.TryGetPropertyValue("inputModalities", out var inputModalities))
        {
            @base.InputModalities = ToStringList(inputModalities);
        }

        if (@base.OutputModalities is null && entry.TryGetPropertyValue("outputModalities", out var outputModalities))
        {
            @base.OutputModalities = ToStringList(outputModalities);
        }

        return @base;
    }

    public static ModelInfo MapModel(JsonNode? raw, string provider)
    {
        var data = raw as JsonObject ?? new JsonObject();
        var info = new ModelInfo
        {
            AdditionalProperties = ToObjectDictionary(data),
        };

        switch (provider)
        {
            case "anthropic":
                SetString(data, "id", value => info.Id = value);
                SetString(data, "display_name", value => info.DisplayName = value);
                info.OwnedBy = "anthropic";
                SetInt(data, "context_length", value => info.ContextWindow = value);
                SetStringList(data, "input_modalities", value => info.InputModalities = value);
                SetStringList(data, "output_modalities", value => info.OutputModalities = value);
                break;

            case "foundry":
                MapFoundryModel(data, info);
                break;

            default:
                SetString(data, "id", value => info.Id = value);
                SetString(data, "owned_by", value => info.OwnedBy = value);
                break;
        }

        if (info.Id is null)
        {
            info.Id = string.Empty;
        }

        return info;
    }

    private static void MapFoundryModel(JsonObject data, ModelInfo info)
    {
        if (data.TryGetPropertyValue("properties", out var propertiesNode) && propertiesNode is JsonObject properties)
        {
            var model = properties["model"] as JsonObject;
            var capabilities = properties["capabilities"] as JsonObject;

            SetString(data, "name", value => info.Id = value);
            SetString(model, "name", value => info.DisplayName = value);
            SetString(model, "publisher", value => info.OwnedBy = value);
            SetInt(model, "maxContextLength", value => info.ContextWindow = value);
            SetStringList(capabilities, "supportedInputModalities", value => info.InputModalities = value);
            SetStringList(capabilities, "supportedOutputModalities", value => info.OutputModalities = value);
            return;
        }

        if (data.ContainsKey("modelName") || StringEquals(data, "type", "ModelDeployment"))
        {
            SetString(data, "name", value => info.Id = value);
            SetString(data, "modelName", value => info.DisplayName = value);
            SetString(data, "modelPublisher", value => info.OwnedBy = value);
            SetInt(data, "maxContextLength", value => info.ContextWindow = value);
            return;
        }

        SetString(data, "id", value => info.Id = value);
        SetString(data, "owned_by", value => info.OwnedBy = value);
        SetInt(data, "maxContextLength", value => info.ContextWindow = value);
    }

    private static JsonObject? MatchCapabilities(string modelId, string provider)
    {
        var providers = CapabilityData.Value["providers"] as JsonObject;
        if (providers?[provider] is not JsonArray entries)
        {
            return null;
        }

        JsonObject? best = null;
        var bestLength = -1;
        foreach (var entryNode in entries)
        {
            if (entryNode is not JsonObject entry)
            {
                continue;
            }

            var prefix = (string?)entry["prefix"];
            if (string.IsNullOrEmpty(prefix) || !MatchesTokenBoundary(modelId, prefix))
            {
                continue;
            }

            if (prefix.Length > bestLength)
            {
                best = entry;
                bestLength = prefix.Length;
            }
        }

        return best;
    }

    private static bool MatchesTokenBoundary(string modelId, string prefix)
    {
        if (modelId == prefix)
        {
            return true;
        }

        return modelId.StartsWith(prefix, StringComparison.Ordinal) &&
            modelId.Length > prefix.Length &&
            !char.IsLetterOrDigit(modelId[prefix.Length]);
    }

    private static JsonObject LoadCapabilityData()
    {
        var assembly = typeof(Discovery).Assembly;
        using var stream = assembly.GetManifestResourceStream(CapabilityResourceName)
            ?? throw new InvalidOperationException($"Embedded resource '{CapabilityResourceName}' was not found.");
        using var reader = new StreamReader(stream);
        return JsonNode.Parse(reader.ReadToEnd()) as JsonObject
            ?? throw new InvalidOperationException("Model capability data must be a JSON object.");
    }

    private static void SetString(JsonObject? data, string key, Action<string> set)
    {
        if (data?.TryGetPropertyValue(key, out var value) == true && value is not null)
        {
            set(value.GetValue<string>());
        }
    }

    private static void SetInt(JsonObject? data, string key, Action<int> set)
    {
        if (data?.TryGetPropertyValue(key, out var value) == true && value is not null)
        {
            set(value.GetValue<int>());
        }
    }

    private static void SetStringList(JsonObject? data, string key, Action<IList<string>> set)
    {
        if (data?.TryGetPropertyValue(key, out var value) == true && value is not null)
        {
            set(ToStringList(value));
        }
    }

    private static bool StringEquals(JsonObject data, string key, string expected) =>
        data.TryGetPropertyValue(key, out var value) &&
        value is not null &&
        string.Equals(value.GetValue<string>(), expected, StringComparison.Ordinal);

    private static IList<string> ToStringList(JsonNode? node)
    {
        if (node is not JsonArray array)
        {
            return [];
        }

        return array
            .Select(item => item?.GetValue<string>() ?? string.Empty)
            .ToList();
    }

    private static Dictionary<string, object?> ToObjectDictionary(JsonObject obj) =>
        obj.ToDictionary(kvp => kvp.Key, kvp => ToObject(kvp.Value));

    private static object? ToObject(JsonNode? node)
    {
        return node switch
        {
            null => null,
            JsonObject obj => ToObjectDictionary(obj),
            JsonArray arr => arr.Select(ToObject).ToList(),
            JsonValue value => ToPrimitive(value),
            _ => null,
        };
    }

    private static object? ToPrimitive(JsonValue value)
    {
        if (value.TryGetValue<string>(out var stringValue))
        {
            return stringValue;
        }

        if (value.TryGetValue<bool>(out var boolValue))
        {
            return boolValue;
        }

        if (value.TryGetValue<int>(out var intValue))
        {
            return intValue;
        }

        if (value.TryGetValue<long>(out var longValue))
        {
            return longValue;
        }

        if (value.TryGetValue<double>(out var doubleValue))
        {
            return doubleValue;
        }

        return value.GetValue<object?>();
    }
}
